Makefile

```diff
@@ -12,130 +12,219 @@
 # language governing permissions and limitations under the License.
 #
 
-.PHONY: all build-linux clean format check-format docker docker-build lint unit-test vet download-portmap build-docker-test build-metrics docker-metrics metrics-unit-test docker-metrics-test docker-vet
-
-IMAGE   ?= amazon/amazon-k8s-cni
-VERSION ?= $(shell git describe --tags --always --dirty)
+.PHONY: all build-linux check clean format check-format docker docker-build lint \
+		unit-test vet download-portmap build-docker-test build-metrics \
+		docker-metrics metrics-unit-test docker-metrics-test docker-vet
+
+# VERSION is the source revision that executables and images are built from.
+VERSION ?= $(shell git describe --tags --always --dirty || echo "unknown")
+
+# DESTDIR is where distribution output (container images) is placed.
+DESTDIR ?= .
+
+# IMAGE is the primary AWS VPC CNI plugin container image.
+IMAGE ?= amazon/amazon-k8s-cni
+IMAGE_NAME ?= $(IMAGE)$(IMAGE_ARCH_SUFFIX):$(VERSION)
+IMAGE_DIST = $(DESTDIR)/$(subst /,_,$(IMAGE_NAME)).tar.gz
+# METRICS_IMAGE is the CNI metrics publisher sidecar container image.
+METRICS_IMAGE ?= amazon/cni-metrics-helper
+METRICS_IMAGE_NAME ?= $(METRICS_IMAGE)$(IMAGE_NAME_SUFFIX):$(VERSION)
+METRICS_IMAGE_DIST = $(DESTDIR)/$(subst /,_,$(METRICS_IMAGE_NAME)).tar.gz
+# TEST_IMAGE is the testing environment container image.
+TEST_IMAGE = amazon-k8s-cni-test
+TEST_IMAGE_NAME = $(TEST_IMAGE)$(IMAGE_NAME_SUFFIX):$(VERSION)
+# These values derive ARCH and DOCKER_ARCH which are needed by dependencies in
+# image build defaulting to system's architecture when not specified.
+#
+# UNAME_ARCH is the runtime architecture of the building host.
+UNAME_ARCH = $(shell uname -m)
+# ARCH is the target architecture which is being built for.
+#
+# These are pairs of input_arch to derived_arch separated by colons:
+ARCH ?= $(lastword $(subst :, ,$(filter $(UNAME_ARCH):%,x86_64:amd64 aarch64:arm64)))
+# DOCKER_ARCH is the docker specific architecture specifier used for building on
+# multiarch container images.
+DOCKER_ARCH ?= $(lastword $(subst :, ,$(filter $(ARCH):%,amd64:amd64 arm64:arm64v8)))
+# GOLANG_IMAGE is the building golang container image used.
+GOLANG_IMAGE ?= golang:1.13-stretch
+# For the requseted build, these are the set of Go specific build environment
+# variables.
+export GOARCH = $(ARCH)
+export GOOS = linux
+export CGO_ENABLED = 0
+# NOTE: Provided for local toolchains that require explicit module feature flag.
+export GO111MODULE = on
+export GOPROXY ?= direct
+
+# LDFLAGS is the set of flags used when building golang executables.
 LDFLAGS ?= -X main.version=$(VERSION)
+# ALLPKGS is the set of packages provided in source.
+ALLPKGS = $(shell go list ./...)
+# BINS is the set of outputted command executables.
+BINS = $(notdir $(wildcard cmd/*))
```

Comment:
I didn't mean much by saying they're `outputted`, we just as easily say `built` or `generated` here. I'd prefer to update the comment to say `built` as that's what these `BINS` are - the executables that were built.

I think you're right about the nature of this dynamic list: the hardcoded list makes more sense as it stands because the names of the cmds don't generally align with the output names (which maybe they should be renamed to? that's an unrelated point here though!).